allennlp/common/mmap.py

```diff
@@ -0,0 +1,251 @@
+            self._sizes = np.frombuffer(
+                self._bin_buffer, dtype=np.int32, count=self._len, offset=offset
+            )
+            self._pointers = np.frombuffer(
+                self._bin_buffer,
+                dtype=np.int64,
+                count=self._len,
+                offset=offset + self._sizes.nbytes,
+            )
+
+        def __del__(self):
+            self._bin_buffer_mmap._mmap.close()
+            del self._bin_buffer_mmap
+
+        @property
+        def dtype(self):
+            return self._dtype
+
+        @property
+        def sizes(self):
+            return self._sizes
+
+        @lru_cache(maxsize=8)
+        def __getitem__(self, i):
+            return self._pointers[i], self._sizes[i]
+
+        def __len__(self):
+            return self._len
+
+    def __init__(self, path):
+        super().__init__()
+
+        self._path = None
+        self._index = None
+        self._bin_buffer = None
+
+        self._do_init(path)
+
+    def __getstate__(self):
+        return self._path
+
+    def __setstate__(self, state):
+        self._do_init(state)
+
+    def _do_init(self, path):
+        self._path = path
+        self._index = self.Index(index_file_path(self._path))
+
+        _warmup_mmap_file(data_file_path(self._path))
+        self._bin_buffer_mmap = np.memmap(data_file_path(self._path), mode="r", order="C")
+        self._bin_buffer = memoryview(self._bin_buffer_mmap)
+
+    def __del__(self):
+        self._bin_buffer_mmap._mmap.close()
+        del self._bin_buffer_mmap
+        del self._index
+
+    def __len__(self):
+        return len(self._index)
+
+    @lru_cache(maxsize=8)
+    def __getitem__(self, i):
+        ptr, size = self._index[i]
+        np_array = np.frombuffer(self._bin_buffer, dtype=self._index.dtype, count=size, offset=ptr)
+        if self._index.dtype != np.int64:
+            np_array = np_array.astype(np.int64)
+
+        return torch.from_numpy(np_array)
+
+    @property
+    def sizes(self):
+        return self._index.sizes
+
+    @staticmethod
+    def exists(path):
+        return os.path.exists(index_file_path(path)) and os.path.exists(data_file_path(path))
+
+
+class MMapIndexedCacheBuilder(object):
+    def __init__(self, out_file, vocab_size=None):
+        self._data_file = open(out_file, "wb")
+        self._sizes = []
+        self._field_names = None
+        if vocab_size is not None and vocab_size < 65500:
+            self._dtype = np.uint16
+        else:
+            self._dtype = np.int32
+
+    def add_instance(self, instance: Instance):
+        tensor_dict = instance.as_tensor_dict()
+        flattened_dict = self.flatten_dict(tensor_dict)
+        if not self._field_names:
+            self._field_names = list(sorted(flattened_dict.keys()))
+            assert self._field_names
+            # TODO: what if some instances have a different set of field names, i.e missing some, for test instances, we don't have supervision.....
```

Comment:
@epwalsh Do you have any ideas how to handle this?

Reply:
Optimally, we would like to write the names of the fields once, but the set of possible fields is different for train and test.

Reply:
Yea, I think it's fine to write the names of each field with each instance. Should be a relatively small cost.

Reply:
Well, I want to solve it anyway if you don't mind haha.
I'll use a some sort  of enum (that can be updated) for each field name and write the enum index instead of the string itself.